### PR TITLE
Keepalive for out_forward plugin

### PR DESCRIPTION
**Pull request by ganmacs**
**Which issue(s) this PR fixes**: 
Fixes https://github.com/fluent/fluentd/issues/2188

**What this PR does / why we need it**: 

Implemented keepalive feature for out_forward plugin.

**Docs Changes**:

~~Needed. I'll add a detail of this feature to  https://github.com/fluent/fluentd-docs/blob/master/docs/v1.0/out_forward.txt~~
https://github.com/fluent/fluentd-docs/pull/632

**Release Note**: 

No need or use PR title?

@repeatedly Will you please review?

### CODE REVIEW

File: lib/fluent/plugin/out_forward.rb [hunk 1]
```diff
@@ -880,6 +1063,32 @@ def on_read(sock, ri, data)
...
       end
+
+      private
+
+      def connect(host = nil)
+        sock = if @keepalive
+                 @socket_cache.fetch_or { @sender.create_transfer_socket(host || resolved_host, port, @hostname) }
+               else
+                 @log.debug('connect new socket')
+                 @sender.create_transfer_socket(resolved_host, port, @hostname)
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: Is this point also `host || resolved_host`?

File: lib/fluent/plugin/out_forward.rb [hunk 2]
```diff
@@ -538,7 +566,141 @@ def ack_reader
...
+            else
+              @log.warn("Not found key for dec_ref_by_value: #{key}")
+            end
+          end
+        end
+
+        # This method is expected to be called in class which doesn't call #fetch_or
+        def revoke_by_value(val)
+          @mutex.synchronize do
+            sock = @active_socks.detect { |_, v| v.sock == val }
```
[the remark this answers is not in the document]
Comment (ganmacs) [reply lib/fluent/plugin/out_forward.rb]: Yes. and added comment https://github.com/fluent/fluentd/pull/2393/commits/8bffce61a10834c71dcd0788a55edb7a7975f0d4

File: lib/fluent/plugin/out_forward.rb [hunk 3]
```diff
@@ -538,7 +566,141 @@ def ack_reader
...
+              return
+            end
+
+            sock = @inactive_socks.detect { |_, v| v.sock == val }
+            if sock
+              key = sock.first
+              @inactive_socks[key].ref -= 1
+              return
+            else
+              @log.warn("Not found key for dec_ref_by_value: #{key}")
```
[the remark this answers is not in the document]
Comment (ganmacs) [reply lib/fluent/plugin/out_forward.rb]: I expect key is `Thread.current.object_id`. So it already returns object id(and `Thread.current.object_id.name` raises `NoMethodError` I think).

File: lib/fluent/plugin/out_forward.rb (hunk 1 above)
Comment (ganmacs) [reply lib/fluent/plugin/out_forward.rb]: Right https://github.com/fluent/fluentd/pull/2393/commits/44cb30170888223fb452ac5858780c95bcff8858

File: lib/fluent/plugin/out_forward.rb [hunk 4]
```diff
@@ -538,7 +566,142 @@ def ack_reader
...
+              @inactive_socks[key] = @active_socks.delete(key)
+              @inactive_socks[key].ref = 0
+            end
+          end
+        end
+
+        def clear
+          @mutex.synchronize do
+            @inactive_socks.values.each do |s|
+              s.sock.close
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: Adding `rescue nil` is better? Hard to recover broken socket here.

File: lib/fluent/plugin/out_forward.rb [hunk 5]
```diff
@@ -538,7 +566,142 @@ def ack_reader
...
+            end
+            @inactive_socks.clear
+
+            @active_socks.values.each do |s|
+              s.sock.close
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: ditto

File: lib/fluent/plugin/out_forward.rb [hunk 6]
```diff
@@ -538,7 +566,142 @@ def ack_reader
...
+          end
+        end
+
+        def purge_obsolete_socks
+          @mutex.synchronize do
+            @inactive_socks.keys.each do |k|
+              # 0 means sockets stored in this class received all acks
+              if @inactive_socks[k].ref <= 0
+                s = @inactive_socks.delete(k)
+                s.sock.close
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: ditto

File: lib/fluent/plugin/out_forward.rb [hunk 7]
```diff
@@ -672,24 +838,42 @@ def send_data_actual(sock, tag, chunk)
...
-        sock = @sender.create_transfer_socket(resolved_host, port, @hostname)
+        sock = connect
+
         begin
           send_data_actual(sock, tag, chunk)
         rescue
-          sock.close rescue nil
+          if @keepalive
+            @socket_cache.revoke
+          end
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: Following code is needed?
```
else
  sock.close rescue nil
end
```

Comment (ganmacs) [reply lib/fluent/plugin/out_forward.rb]: Thanks. I didn't mean to delete the code. https://github.com/fluent/fluentd/pull/2393/commits/d3cb12013f13c4b849f1359d04ca8bf962b7af7b

**Comment by repeatedly** [comment]
If no reviews from developers, I will merge soon.

**Comment by repeatedly** [comment]
Thanks for hard work!

**Pull request merged**